### DISCUSSION TITLE
Pandas :var replacements in strings.

### DISCUSSION THREAD

**Discussion by Zethtren** [sqlalchemy/sqlalchemy#10976] [Ideas]
Not entirely sure how to phrase the topic. And not sure that its an issue (but would be a nice feature). So I apologize if there is an ongoing discussion for this. I could not find it.

This issue only arose because I have "unclean" data where "Something: other thing" sometimes gets typed in as "Something :other thing".

The end result looks like this. 
```python
query = f"""
REPLACE INTO {table} (`{"`, `".join(list(items.columns))}`)
VALUES (:{", :".join(list(items.columns))})
"""
values = items.replace({np.nan: None}).to_dict(orient="records")
conn.execute(text(query), values)
```

And this works beautifully except for the case above. 

I have solved this with a "janky" regex solution. (Which could be improved to maintain data consistency by escaping the ":" instead of moving the space). But, works as needed for my purposes.

And that solution looks like this. 

```python
sql_rep = re.compile(r"([\s]+:|^:)([^\s]+)")
items.loc[
    :, (items.applymap(type) == str).all(0)
] = items.loc[
    :, (items.applymap(type) == str).all(0)
].applymap(
    lambda x: re.sub(sql_rep, r"\1 \2", x)
)
```

I would think it would be better that the default behavior of execute() only applied the insertions that are present in the text() before values is inserted as opposed to doing them both. This is a breaking change though and maybe a kw arg could exist to enforce this behavior so I no longer need to modify the data to break that pattern.

It might be that this already exists? If so I would love to be pointed in the right direction to remove this eyesore. I am also willing to work with you to add that option myself. (One that only replaces the initial inputs as opposed to the ones added by the string and not some obtuse regex replace on the data itself.)

**Comment by zzzeek** [comment]
Hey there -

I'm not sure what this is about and this might be due to terminology.  "default behavior of execute() only applied the insertions" , what's an "insertion"?   Are you referring to bound parameters and using them with the `text()` construct which uses colons for the syntax?    why would bound parameters be embedded in pandas data?    Overall need a better description of what you are trying to do.

**Reply by Zethtren** [reply]
Yeah, absolutely. 

Sorry for the phrasing and the "bound" parameters aren't there intentionally. They are definitely typos on the input data.

But any time. I have a dataframe.
```python
df = pd.DataFrame({"a": [":fails", "runs:fine", "also: fine", "also :fails"]})
```

When the `execute(text(query), df.to_dict(orient="records"))` is called the query will fail on. "No bind parameter for ":fails". "

Which implies that it replaces them all and then looks to replace them all again.

**Reply by zzzeek** [reply]
OK those are colons which you have to escape if you are using them in [text()](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text):

```
stmt = text(my_pandas_string.replace(":", "\:"))
conn.execute(stmt)
```

**Reply by Zethtren** [reply]
```python
import pandas as pd
from sqlalchemy import text
from sqlalchemy.dialects import mysql

df = pd.DataFrame({"a": [":fails", "runs:fine", "also: fine", "also :fails"]})

query = f"""
REPLACE INTO test_table (`{"`, `".join(list(df.columns))}`)
VALUES (:{", :".join(list(df.columns))})
"""

print(text(query).compile(dialect=mysql.dialect()))
```

We're mis-communicating I think. So this is the result from what does into text. 

It ends up with 
VALUES( %s, %s, ...) etc for each column.

but when I do execute(text(query), df.to_dict(orient="records"))

what is return should be more like this.

```python
"""
REPLACE INTO test_table (`a`)
VALUES (":fails"), ("runs:fine"), ("also: fine"), ("also, :fails")
"""
```

And this works EXACTLY like that when there is no " :var" in the string. 
However that example above will tell me there is no bind parameter for `:fails`.

It works perfectly fine otherwise.

**Reply by zzzeek** [reply]
yes, add replace as shown

```py
import pandas as pd
from sqlalchemy import text
from sqlalchemy.dialects import mysql

df = pd.DataFrame({"a": [":fails", "runs:fine", "also: fine", "also :fails"]})

query = f"""
REPLACE INTO test_table (`{"`, `".join(list(df.columns))}`)
VALUES (:{", :".join(list(df.columns))})
"""

print(text(query.replace(":", "\:")).compile(dialect=mysql.dialect()))
```

**Reply by Zethtren** [reply]
Is there a way that I can see what.

`conn.execute(text(query.replace(":", "\:")), df....)` looks like? Just to confirm the behavior?
without executing.

**Reply by zzzeek** [reply]
```
print(query.replace(":", "\:"))
```

will get you that

**Reply by CaselIT** [reply]
that does not seem correct. I the query the user wants to execute is
```sql
REPLACE INTO test_table (`a`)
VALUES (:a)
```

that converted into text finds a single bind parameter named `a`. At that point a list of dict with a key a should work correctly, independently of any data that's in the values

@Zethtren can you provide a runnable example of failing code?

**Reply by Zethtren** [reply]
I have been trying to get a working sample to showcase the problem without having to execute but I can't get it to print the way that it runs against the DB.

This is what I have so far. 

```python
import pandas as pd
from sqlalchemy import text, bindparam
from sqlalchemy.dialects import mysql


df_fails = pd.DataFrame(
    {"a": [":other", "fine:other", "fine: asdf", "another :badone"]}
)
df_works = pd.DataFrame({"a": ["other", "fine:other", "fine: asdf", "another"]})

query = f"""
REPLACE INTO test_table (`{"`, `".join(list(df_works.columns))}`)
VALUES (:{", :".join(list(df_works.columns))})
"""

# Sameple bind trying to get a printed version because execute lets me bind differently.
sample = [
    ":other",
    "fine:other",
    "fine: asdf",
    "another :badone",
]

# NOTE: This fails because "a" is no longer bindable
# comp1 = (
#     text(query.replace(":", "\:"))
#     .bindparams(bindparam(key="a", value=sample, expanding=True))
#     .compile(
#         compile_kwargs={"literal_binds": True},
#         dialect=mysql.dialect(),
#     )
# )
# print(comp1)

# NOTE: This gives the result I want exactly.
# But when handled through execute() the behavior is different.
comp2 = (
    text(query)
    .bindparams(bindparam(key="a", value=sample, expanding=True))
    .compile(
        dialect=mysql.dialect(),
        compile_kwargs={"literal_binds": True},
    )
)
print(comp2)
print(comp2.statement.compile(dialect=mysql.dialect()))


# NOTE: Commented out because I don't want to build a whole connection.
# # DBCONN
# # This one will work as expected if tied to a MySQL connection.
# conn.execute(text(query), df_works.to_dict(orient="records"))
#
# # DBCONN
# # This one will fail because it will try to bind the values inside the replaced strings as well
# conn.execute(text(query), df_fails.to_dict(orient="records"))


# Ideally I don't have to modify all of the dataframe data to escape every ":"
# I just don't understand why the replacement seems to recurse instead of doing the whole thing.
```

**Reply by Zethtren** [reply]
The main problem is that this is extremely dynamic so I can't reasonably construct something like comp2.compile in this code every time without an absolute boat of boilerplate code and regex replacing the data is far more efficient. 

It was just a tricky but to track down because it was trying to bind the strings :badone and :other after the initial compilation using the bottom pattern with the conn.execute()

**Reply by CaselIT** [reply]
> Ideally I don't have to modify all of the dataframe data to escape every ":"

this is should not be necessary. Only the `:` that aren't bindparam in the query need escaping, not the one in the data provided.

I'll take a look

**Reply by Zethtren** [reply]
@CaselIT I can't seem to replicate it on sqlite. And the sample I have can't be shared because of NDA and BAAs. However, I can share that they are relatively long comment strings with new lines, tabs, etc and many characters. So maybe that behaves differently than SQLite? Or there are other bindings in the MySQL version that could be attributed to it. I will work through the failing case that I have to and try to replicate it on SQLLite. 

But, basically it looks like this:
```python
import pandas as pd
from sqlalchemy import text, bindparam, create_engine
from sqlalchemy.dialects import mysql
from pathlib import Path

p = Path(".")
with (p / "test.db").open("w"):
    pass


engine = create_engine("sqlite:///test.db", echo=True)

df_fails = pd.DataFrame(
    {
        "a": [":other", "fine:other", "fine: asdf", "another :badone"],
        "b": [1, 2, 3, 4],
    }
)
df_works = pd.DataFrame(
    {"a": ["other", "fine:other", "fine: asdf", "another"], "b": [1, 2, 3, 4]}
)

with engine.begin() as conn:
    conn.execute(text("CREATE TABLE IF NOT EXISTS test_table (a VARCHAR, b INT)"))

query = f"""
INSERT INTO test_table (`{"`, `".join(list(df_works.columns))}`)
VALUES (:{", :".join(list(df_works.columns))})
"""
with engine.begin() as conn:
    conn.execute(text(query), df_works.to_dict(orient="records"))

with engine.begin() as conn:
    conn.execute(text(query), df_fails.to_dict(orient="records"))
```

Except the 2nd one is failing stating a bind parameter is needed for :badone. 

Again it happens everytime on the MySQL data but, has not here.
I will see if I can get you a better sample (One that actually fails).

**Reply by Zethtren** [reply]
I do see "\u202f" characters in the slightly cleaned up version that I am working with. 

I will try spinning up a mysql db locally and replicating it. 

`Failed insert with StatementError("(sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'receptive'")`

This is the exact error I got in prod on the code that was interfacing with mysql. 

replacing ":receptive" with ": receptive" using the regex above fixes it but the behavior for creating the query is identical to the samples above. I will look into some other things on the pandas reading the data side as well as intermediate storage etc.

**Reply by CaselIT** [reply]
can you verify that the `:receptive` is not in the column name instead of the values?

**Comment by CaselIT** [comment]
Trying what you provided. I can't reproduce your errors.

The only reason what may cause an error that I can thing of is if the a *column name* in the dataframe has a colon in it, not if the values have it. Please verify it that's the case

Try the script below to see if you can reproduce the error:
```py
import pandas as pd
from sqlalchemy import text
import sqlalchemy as sa


for url in (
    "mysql+pymysql://scott:tiger@127.0.0.1:3306/test?charset=utf8mb4",
    "sqlite://",
):
    engine = sa.create_engine(url, echo=True)
    df_fails = pd.DataFrame(
        {"a": [":other", "fine:other", "fine: asdf", "another :badone"]}
    )
    df_works = pd.DataFrame(
        {"a": ["other", "fine:other", "fine: asdf", "another"]}
    )
    query = f"""
    REPLACE INTO test_table (`{"`, `".join(list(df_works.columns))}`)
    VALUES (:{", :".join(list(df_works.columns))})
    """
    print("using", engine.dialect.name)
    with engine.connect() as conn:
        conn.exec_driver_sql("drop table if exists test_table ")
        conn.exec_driver_sql("create table test_table(a varchar(4096))")
        conn.execute(text(query), df_works.to_dict(orient="records"))
        conn.execute(text(query), df_fails.to_dict(orient="records"))

    df_fails = pd.DataFrame(
        {
            "a": [":other", "fine:other", "fine: asdf", "another :badone"],
            "b": [1, 2, 3, 4],
        }
    )
    df_works = pd.DataFrame(
        {
            "a": ["other", "fine:other", "fine: asdf", "another"],
            "b": [1, 2, 3, 4],
        }
    )

    with engine.begin() as conn:
        conn.exec_driver_sql("drop table if exists test_table2 ")
        conn.exec_driver_sql(
            "CREATE TABLE IF NOT EXISTS test_table2 (a VARCHAR(4096), b INT)"
        )

    query = f"""
    INSERT INTO test_table2 (`{"`, `".join(list(df_works.columns))}`)
    VALUES (:{", :".join(list(df_works.columns))})
    """
    with engine.begin() as conn:
        conn.execute(text(query), df_works.to_dict(orient="records"))

    with engine.begin() as conn:
        conn.execute(text(query), df_fails.to_dict(orient="records"))

```

**Reply by Zethtren** [reply]
Nope. No colons in Column names. They are all PascalCasedStrings. 

Does adding "\r\n \u202f\r\n :fails" as one of the samples affect anything? Out of quick curiosity analyzing the value that broke on my end. 

If not that's fine. I will dig into this. It was consistently failing on the production system and was immediately fixed by the regex correction. So I know there is something valid there. I'm just not entirely sure how to reproduce it. I can say that I am pulling the string (represent with the \r\n \u202f etc) from MS SQL into Pandas and then running this process. So maybe its a pandas issue and the data is being altered internally. Either way it won't be quick for me to set up tests on this right now. I can re-open / open a new discussion when I have a reproducible sample that I can paste to you.

**Reply by Zethtren** [reply]
OH! The other thing is that its going to be "too big" to run on a single insert so it gets batched in how its sent. If the value is split into a second batch maybe there is some compile behavior that causes it to read that :var again?

**Reply by Zethtren** [reply]
That's not a behavior I defined. I executed it exactly as above. I just remember (perhaps falsely) reading that it batches queries beyond a certain size (or streams them in or something different?)

**Reply by CaselIT** [reply]
> Nope. No colons in Column names. They are all PascalCasedStrings.

that's the only way the example above should give you an error like what you are seeing. Please try to modify it to be closer to your use case.

> Does adding "\r\n \u202f\r\n :fails" as one of the samples affect anything? Out of quick curiosity analyzing the value that broke on my end.

Nope everything keep working. 

From what you are describing it seems that you are not doing the following
```py
query = f"""
    REPLACE INTO test_table (`{"`, `".join(list(df_works.columns))}`)
    VALUES (:{", :".join(list(df_works.columns))})
    """
conn.execute(text(query), df_works.to_dict(orient="records"))
```
but something like this
```py
query = f"""
    REPLACE INTO test_table (`{"`, `".join(list(df_works.columns))}`)
    VALUES {",".join('(' + ', '.join(
      str(val) for row in df_works.to_dict(orient="records") for val in row.values()
    ) + ')'}
    """ 
conn.execute(text(query))
```
meaning that you are trying to construct the values yourself in the query, not use bound parameters

**Reply by Zethtren** [reply]
Mmm no, I definitely have (and had before applying the regex fix) the first format. The issue occurs at the execute block as pasted above. But maybe it's some kind of character encoding thing reading the values in from MS SQL. I will stand up a dummy on the known failing case and try to narrow down the failing bit before bringing it up again. 

Thanks for the support!

**Reply by CaselIT** [reply]
Ok sounds good!
